CHANGE: Removed aliased values in Gamepad enum. (ISXB-543)

adrian-koretski-unity3d opened the pull request:
### Description

Fixing confusing menu when selecting face buttons due to aliased enum values. 

### Changes made

Removed aliased enum values. 

### Testing

Manual testing. 

### Risk

None. 

### Checklist

Before review:

- [ ] Changelog entry added.
    - Explains the change in `Changed`, `Fixed`, `Added` sections.
    - For API change contains an example snippet and/or migration example.
    - JIRA ticket linked, example ([case %<ID>%](https://issuetracker.unity3d.com/product/unity/issues/guid/<ID>)). If it is a private issue, just add the case ID without a link.
    - Jira port for the next release set as "Resolved".
- [ ] Tests added/changed, if applicable.
    - Functional tests `Area_CanDoX`, `Area_CanDoX_EvenIfYIsTheCase`, `Area_WhenIDoX_AndYHappens_ThisIsTheResult`.
    - Performance tests.
    - Integration tests.
- [ ] Docs for new/changed API's.
    - Xmldoc cross references are set correctly.
    - Added explanation how the API works.
    - Usage code examples added.
    - The manual is updated, if needed.

During merge:

- [ ] Commit message for squash-merge is prefixed with one of the list:
    - `NEW: ___`.
    - `FIX: ___`.
    - `DOCS: ___`.
    - `CHANGE: ___`.
    - `RELEASE: 1.1.0-preview.3`.

After merge:

- [ ] Create forward/backward port if needed. If you are blocked from creating a forward port now please add a task to ISX-1444.

adrian-koretski-unity3d committed:
ISXB-543   Implemented custom drawer base for enums as well as on specific to gamepad buttons. Also removed aliased values in GamepadButton enum

adrian-koretski-unity3d committed:
Removed custom enum drawer.

adrian-koretski-unity3d committed:
Added changes to changelog.

unity-cla-assistant commented:
[![CLA assistant check](https://cla-assistant.cds.internal.unity3d.com/pull/badge/signed)](https://cla-assistant.cds.internal.unity3d.com/Unity-Technologies/InputSystem?pullRequest=1983) <br/>All committers have signed the CLA.

adrian-koretski-unity3d committed:
Merge branch 'develop' into isxb-543

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Devices/Gamepad.cs:
```diff
@@ -326,64 +326,6 @@ public enum GamepadButton
         /// The right trigger button on a gamepad.
         /// </summary>
         RightTrigger = 33,
-
```
We cannot remove these from this enum without requiring a major version revision since it would break public API for existing customers (and according to semver), hence this is not a viable way forward.

adrian-koretski-unity3d replied:
I hadn't thought of this. I re-added the aliased enum values (although I did not reset them in the other files that used them to move towards north/south/east/west). I also added a slightly modified version of the aliased enum drawer from [#1862](https://github.com/Unity-Technologies/InputSystem/pull/1862) where we can remove values from the drawer.

ekcoh commented on Packages/com.unity.inputsystem/CHANGELOG.md:
```diff
@@ -19,6 +19,7 @@ however, it has to be formatted properly to pass verification tests.
 
 ### Changed
 - Use `ProfilerMarker` instead of `Profiler.BeginSample` and `Profiler.EndSample` when appropriate to enable recording of profiling data.
+- Removed aliased values in GamepadButton enum. 
```
We do not want to release a major revision due to this change.

adrian-koretski-unity3d replied:
Should we plan to do so down the line? Perhaps mark the aliased values as depricated (I'm not sure how we go about doing this).

ekcoh replied:
We basically shouldn't modify the enum at all, we should aim for a completely editor-based "fix" here. There is no functional issue with the current solution, its only confusing users from UX perspective.

ekcoh replied:
It makes sense though to file an internal ticket to consider removing aliases for control enums down the line (in a major future release). I have now filed it.

ekcoh commented:
I would recommend considering a fix based off this one but with inlined/combined options for aliased entries? https://github.com/Unity-Technologies/InputSystem/pull/1862

ekcoh commented:
Seems like @Billreyn  was also considering that unless opinion has shifted, see https://github.com/Unity-Technologies/InputSystem/pull/1862#issuecomment-1971097297

adrian-koretski-unity3d committed:
Re-added aliased enums into GamepadButton for compatibility with existing projects. Added custom aliased enum drawer base and custom enum drawer for GamepadButton to hide aliased values.

adrian-koretski-unity3d commented:
> Seems like @Billreyn was also considering that unless opinion has shifted, see [#1862 (comment)](https://github.com/Unity-Technologies/InputSystem/pull/1862#issuecomment-1971097297)

I believe so. Based on our conversation, we want to display only North/South/East/West

adrian-koretski-unity3d committed:
Merge branch 'develop' of https://github.com/Unity-Technologies/InputSystem into isxb-543

adrian-koretski-unity3d committed:
Ran the formatter on commit changes.

adrian-koretski-unity3d committed:
Merge branch 'develop' into isxb-543

adrian-koretski-unity3d committed:
Merge branch 'develop' into isxb-543

ekcoh commented on Assets/Tests/InputSystem/CoreTests_Actions.cs:
```diff
@@ -4661,7 +4661,7 @@ public void Actions_PressingAndReleasingButtonInSameUpdate_StillTriggersAction()
             ctx => { ++receivedCalls; };
         action.Enable();
 
-        var firstState = new GamepadState {buttons = 1 << (int)GamepadButton.B};
```
The fix shouldn't have any impact on existing tests?

ekcoh replied:
Hence I wouldn't expect any changes to any tests.

adrian-koretski-unity3d replied:
I believe this is due to my first pass where I removed the aliased values. Will revert

ekcoh commented on Assets/Tests/InputSystem/Plugins/UserTests.cs:
```diff
@@ -942,7 +942,7 @@ public void Users_CanDetectUseOfUnpairedDevice()
 
         ++InputUser.listenForUnpairedDeviceActivity;
 
-        InputSystem.QueueStateEvent(gamepad, new GamepadState { leftStick = new Vector2(1, 0)}.WithButton(GamepadButton.A));
```
This shouldn't need to change either.

adrian-koretski-unity3d replied:
I believe this is due to my first pass where I removed the aliased values. Will revert

ekcoh commented:
> > Seems like @Billreyn was also considering that unless opinion has shifted, see [#1862 (comment)](https://github.com/Unity-Technologies/InputSystem/pull/1862#issuecomment-1971097297)
> 
> I believe so. Based on our conversation, we want to display only North/South/East/West

Exactly, with emphasis on "display", hence it should be property drawer only changes as I see it.

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Devices/Gamepad.cs:
```diff
@@ -255,31 +255,31 @@ public enum GamepadButton
         /// </summary>
         /// <remarks>
-        /// Identical to <see cref="Y"/> and <see cref="Triangle"/> which are the Xbox and PlayStation controller names for this button.
+        /// Identical to <see cref="North"/> and <see cref="North"/> which are the Xbox and PlayStation controller names for this button.
```
This seems incorrect? Referencing to the same thing

adrian-koretski-unity3d replied:
I believe this is due to my first pass where I removed the aliased values. Will revert

ekcoh commented on Assets/Tests/InputSystem/CoreTests_Events.cs:
```diff
@@ -103,7 +103,7 @@ public void Events_CanListenForEvents()
         allButtonPresses.Clear();
 
         Release(gamepad.buttonSouth);
-        InputSystem.QueueStateEvent(gamepad, new GamepadState(GamepadButton.A, GamepadButton.B));
```
Not expecting any changes needed to this file

ekcoh commented on Assets/Tests/InputSystem/CoreTests_State.cs:
```diff
@@ -424,7 +424,7 @@ public void State_CanUpdateButtonStateUsingEvent()
 
         Assert.That(gamepad.buttonEast.isPressed, Is.False);
 
-        var newState = new GamepadState {buttons = 1 << (int)GamepadButton.B};
```
Not expecting any changes needed to this file

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Editor/PropertyDrawers/AliasedEnumPropertyDrawer.cs:
```diff
@@ -0,0 +1,72 @@
+#if UNITY_EDITOR
```
![Screenshot 2024-09-13 at 11 09 14](https://github.com/user-attachments/assets/3fc87754-2d72-4119-a846-f2b9561ff459)
![Screenshot 2024-09-13 at 11 11 23](https://github.com/user-attachments/assets/6b477234-d04b-45e0-b162-5f723599c821)
I am afraid this is not doing what its supposed to.....

The following things need to be addressed:
- Aliased values should be part of the same entry when the inspector is displayed, e.g. "South / X / A"
- Drop down need to allow all values to be picked, e.g. either "South / X / A" or each one individually would be fine from my perspective.
- At the moment when selecting e.g. South, the value displayed is "North", so something with mapping is incorrect

adrian-koretski-unity3d replied:
I'll go over it and make sure it behaves and displays correctly

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Editor/PropertyDrawers/AliasedEnumPropertyDrawer.cs:
```diff
@@ -0,0 +1,72 @@
+
+        private void ProcessDisplayNamesForAliasedEnums()
+        {
+            var enumNamesAndValues = new Dictionary<string, int>();
```
Since the enum is small and we want consistency, maybe skip this base class and just move into specific GamepadButtonPropertyDrawer class and just use a big switch statement to get it right? That would be quite straightforward. Otherwise there needs to be some smartness to get the order of each aliased combination separated by slashes consistent.

ekcoh reviewed (changes requested):
Needs a bit more work I believe

adrian-koretski-unity3d committed:
Revert "ISXB-543   Implemented custom drawer base for enums as well as on specific to gamepad buttons. Also removed aliased values in GamepadButton enum"

This reverts commit 110e5c6df1671dc258a083a87b2c5bceb47d40c0.

# Conflicts:
#	Packages/com.unity.inputsystem/InputSystem/Editor/PropertyDrawers/AliasedEnumPropertyDrawer.cs.meta
#	Packages/com.unity.inputsystem/InputSystem/Editor/PropertyDrawers/GpadButtonPropertyDrawer.cs.meta

adrian-koretski-unity3d committed:
Merge branch 'isxb-543' of https://github.com/Unity-Technologies/InputSystem into isxb-543

adrian-koretski-unity3d committed:
Reworked GpadButtonPropertyDrawer according to feedback.

Removed general AliasedEnumPropertyDrawer script.

lyndon-unity commented on Packages/com.unity.inputsystem/InputSystem/Editor/PropertyDrawers/GpadButtonPropertyDrawer.cs:
```diff
@@ -0,0 +1,82 @@
+    /// Property drawer for <see cref = "GamepadButton" />
+    /// </summary >
+    [CustomPropertyDrawer(typeof(GamepadButton))]
+    internal class GpadButtonPropertyDrawer : PropertyDrawer
```
Minor but can you use a full name here rather than abbreviation 
GpadButtonPropertyDrawer ->
GamepadButtonPropertyDrawer

ekcoh replied:
Agree

ekcoh replied:
Which also means renaming file and .meta

lyndon-unity reviewed (approved):
Code looks reasonable, one minor recommendation, but I haven't tested it in action.
Suggest adding QA or design for a manual verification

ekcoh commented on Packages/com.unity.inputsystem/CHANGELOG.md:
```diff
@@ -27,6 +27,7 @@ however, it has to be formatted properly to pass verification tests.
 
 ### Changed
 - Use `ProfilerMarker` instead of `Profiler.BeginSample` and `Profiler.EndSample` when appropriate to enable recording of profiling data.
+- Removed aliased values in GamepadButton enum.
```
This should be reverted as well, suggest to change to something like e.g. "Changed representation of GamepadButton enum values in Inspector to display aliased enum values as a single item to avoid confusion around selection and aliased value display when multiple enum items map to the same numerical value."

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Editor/PropertyDrawers/GpadButtonPropertyDrawer.cs:
```diff
@@ -0,0 +1,82 @@
+
+            for (var i = 0; i < enumDisplayNames.Length; ++i)
+            {
+                string enumName = enumDisplayNames[i];
```
This assignment could be on line 40 right since overwritten for the special cases?

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Editor/PropertyDrawers/GpadButtonPropertyDrawer.cs:
```diff
@@ -0,0 +1,82 @@
+
+                if (enumName != null)
+                {
+                    enumNamesAndValues.Add(enumName, (int)enumValues.GetValue(i));
```
Could be moved up to the place where it is currently assigned as null?

adrian-koretski-unity3d replied:
I changed it up a little bit. For the cases where we do not want to enter the enum into the list, I continue the for loop, eliminating the need for null/null check. The switch statement sets an string value which is then entered into the list.

ekcoh reviewed (approved):
Approving this to not block this PR in any way while I have limited accessability. I didn't run it on this review round so recommend checking functionality in QA pass.

adrian-koretski-unity3d committed:
Merge branch 'develop' into isxb-543

Pauliusd01 commented:
Will check this today or tomorrow morning at the latest

Pauliusd01 reviewed (changes requested):
Entering play mode with one of these new values selected breaks inspector and throws errors. Also, the enum value does not match the one in the inspector (See debug log in screenshot, whether that's expected I dunno) 

![1726666305 IN-45621_-_Scene_-_Windows,_Mac,_Linux_-_Unity_6_P](https://github.com/user-attachments/assets/ee76ba14-77c2-4ce8-8baf-155a29914106)

adrian-koretski-unity3d committed:
Fixed GamepadButtonPropertyDrawer issues based on PR feedback.

-Renamed GpadButtonPropertyDrawer to GamepadButtonPropertyDrawer.
-Fixed nullref issue caused by m_EnumDisplayNames being reset when entering/exiting play mode.
-Simplified code for setting up enum list.
-Updated CHANGELOG.md

adrian-koretski-unity3d commented:
> Entering play mode with one of these new values selected breaks inspector and throws errors. Also, the enum value does not match the one in the inspector (See debug log in screenshot, whether that's expected I dunno)
> 
> ![1726666305 IN-45621_-Scene-_Windows,_Mac,Linux-_Unity_6_P](https://private-user-images.githubusercontent.com/54306142/368590123-ee76ba14-77c2-4ce8-8baf-155a29914106.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjY2NjY5NzUsIm5iZiI6MTcyNjY2NjY3NSwicGF0aCI6Ii81NDMwNjE0Mi8zNjg1OTAxMjMtZWU3NmJhMTQtNzdjMi00Y2U4LThiYWYtMTU1YTI5OTE0MTA2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTE4VDEzMzc1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc1M2JmMTUzNGU1MTc0NjU3MWViYzdlNzQxZDczZDY0OWMzYWYxOGZmMjY0YjZjODEwNDA4NzdmNTg4Mzc5MjAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.aggLVmzkuk6Qz6cx0KiqRH8r7ZKNNslBhBSh5Wkjgfc)

The nullref issue should be fixed now. The value being displayed will be one of the aliased enums for that value so this is to be expected.

Pauliusd01 commented on Packages/com.unity.inputsystem/CHANGELOG.md:
```diff
@@ -32,6 +32,7 @@ however, it has to be formatted properly to pass verification tests.
 
 ### Changed
 - Use `ProfilerMarker` instead of `Profiler.BeginSample` and `Profiler.EndSample` when appropriate to enable recording of profiling data.
+- Changed representation of GamepadButton enum values in Inspector to display aliased enum values as a single item to avoid confusion around selection and aliased value display when multiple enum items map to the same numerical value.
```
Changelog entry in the wrong version and no bug issue link

adrian-koretski-unity3d committed:
Moved changelog entry for ISXB-543 into correct version and added link to jira.

Pauliusd01 commented:
@adrian-koretski-unity3d I approved too hastily, the link has to be to a public issue tracker not our internal jira lol

adrian-koretski-unity3d committed:
Changed jira link in changelog for ISXB-543 to public issue tracker.
